casper/src/main/scala/coop/rchain/casper/util/BondsParser.scala

```diff
@@ -1,113 +1,132 @@
-      .use(_.getLines)
-      .flatTap(
-        bondsStrs =>
-          Sync[F]
-            .raiseError(
-              new Exception(
-                s"BONDS FILE $bondsPath IS EMPTY. Please fill it or remove so populated bonds file is created on startup."
-              )
-            )
-            .whenA(bondsStrs.isEmpty)
-      )
-      .map { lines =>
-        Try {
-          lines
-            .map(line => {
-              val Array(pk, stake) = line.trim.split(" ")
-              PublicKey(Base16.unsafeDecode(pk)) -> stake.toLong
-            })
-            .toMap
-        }
-      }
-      .flatMap {
-        case Success(bonds) =>
-          bonds.toList
-            .traverse_ {
-              case (pk, stake) =>
-                Log[F].info(s"Bond loaded ${Base16.encode(pk.bytes)} => $stake")
-            }
-            .as(bonds)
-        case Failure(e) =>
-          Sync[F].raiseError(
-            new Exception(s"FAILED PARSING BONDS FILE $bondsPath: ${e}")
+  /**
+    * Parser for bonds file used in genesis ceremony to set initial validators.
+    *
+    * TODO: Create Blocker scheduler for file operations. For now it's ok because it's used only once at genesis.
+    *   Cats Effect 3 removed ContextShift and Blocker.
+    *     - https://typelevel.org/cats-effect/docs/migration-guide#blocker
+    */
+  def parse[F[_]: Sync: ContextShift: Log](bondsPath: Path): F[Map[PublicKey, Long]] = {
+    def readLines(blocker: Blocker) =
+      io.file
+        .readAll[F](bondsPath, blocker, chunkSize = 4096)
+        .through(text.utf8Decode)
+        .through(text.lines)
+        .filter(_.trim.nonEmpty)
+        .evalMap { line =>
+          val lineFormat = "<public_key> <stake>"
+          val lineRegex  = raw"^([0-9a-fA-F]+) ([0-9]+)".r.unanchored
+
+          // Line parser
+          val pubKeyAndStake = tryWithMsg {
+            line match { case lineRegex(fst, snd, _*) => (fst, snd) }
+          }(failMsg = s"INVALID LINE FORMAT: `$lineFormat`, actual: `$line``")
+
+          // Public key parser
+          def publicKey(publicKeyStr: String) =
+            Base16
+              .decode(publicKeyStr)
+              .map(PublicKey(_))
+              .liftTo[F](new Exception(s"INVALID PUBLIC KEY: `$publicKeyStr`"))
+
+          // Stake parser
+          def stake(stakeStr: String) = tryWithMsg(stakeStr.toLong)(
+            failMsg = s"INVALID STAKE `$stakeStr`. Please put positive number."
           )
-      }
 
-  def parse[F[_]: Sync: Log: RaiseIOError](
+          // Parse public key and stake
+          pubKeyAndStake.flatMap(_.bitraverse(publicKey, stake))
+        }
+        .evalTap {
+          case (pk, stake) => Log[F].info(s"Bond loaded ${Base16.encode(pk.bytes)} => $stake")
+        }
+        .compile
+        .to(Map)
+        .adaptErr {
+          case ex: Throwable =>
+            new Exception(s"FAILED PARSING BONDS FILE: $bondsPath\n$ex")
+        }
+    Blocker[F].use(readLines)
```

Comment:
Question.
In cat-effect 2, `Blocker[F].use` would submit this job into a special thread pools which handle blocking IO?

Reply:
Yes, `Blocker[F]` creates a new scheduler with infinite thread pool and its threads are used inside `Blocker[F].use`. I didn't want to create instance on object level so that after the call it will be disposed with GC. 
https://github.com/typelevel/cats-effect/blob/v2.3.1/core/jvm/src/main/scala/cats/effect/internals/BlockerPlatform.scala#L31-L38

In my test of RSpace with blocker, monix io scheduler is created in main method. `Sync.blocking` will be more convenient.
https://github.com/rchain/rchain/blob/42e97c17578082ec7436069c1cf1d3a8417e99d0/node/src/main/scala/coop/rchain/node/Main.scala#L52-L54

https://github.com/rchain/rchain/pull/3441/files#diff-6862710cfeaf5d850532b0e40e2eb9de55451c1a9992c9c4ef1d17f379ee656fR52-R54

Reply:
Why not run it on main scheduler I'm curious?

Reply:
Running on main scheduler (which has fixed number of Java thread reserved) means that we _loose_ a thread blocked even though CPU is not used at all because we are waiting on disk IO to return and thread continues. Monix cannot _reuse_ the thread because there is no _flatMap_ to suspend the Task as light abstraction above Java thread.

Reply:
I see, I mean - this code is something that is run very rarely, when genesis is created. So as I get this is used here to make it proper way whether it is used a lot or not.